autopush-common/src/db/mod.rs

```diff
@@ -1,594 +1,427 @@
-    ) -> impl Future<Item = HelloResponse, Error = ApcError> {
-        trace!(
-            "### uaid {:?}, defer_registration: {:?}",
-            &uaid,
-            &defer_registration
-        );
-        let response: MyFuture<(HelloResponse, Option<DynamoDbUser>)> = if let Some(uaid) = uaid {
-            commands::lookup_user(
-                self.ddb.clone(),
-                self.metrics.clone(),
-                uaid,
-                connected_at,
-                router_url,
-                &self.router_table_name,
-                &self.message_table_names,
-                &self.current_message_month,
-            )
-        } else {
-            Box::new(future::ok((
-                HelloResponse {
-                    message_month: self.current_message_month.clone(),
-                    connected_at,
-                    ..Default::default()
-                },
-                None,
-            )))
-        };
-        let ddb = self.ddb.clone();
-        let router_url = router_url.to_string();
-        let router_table_name = self.router_table_name.clone();
-        let connected_at = connected_at;
-
-        response.and_then(move |(mut hello_response, user_opt)| {
-            trace!(
-                "### Hello Response: {:?}, {:?}",
-                hello_response.uaid,
-                user_opt
-            );
-            let hello_message_month = hello_response.message_month.clone();
-            let user = user_opt.unwrap_or_else(|| DynamoDbUser {
-                current_month: Some(hello_message_month),
-                node_id: Some(router_url),
-                connected_at,
-                ..Default::default()
-            });
-            let uaid = user.uaid;
-            trace!("### UAID = {:?}", &uaid);
-            let mut err_response = hello_response.clone();
-            err_response.connected_at = connected_at;
-            if !defer_registration {
-                future::Either::A(
-                    commands::register_user(ddb, &user, &router_table_name)
-                        .and_then(move |result| {
-                            debug!("Success adding user, item output: {:?}", result);
-                            hello_response.uaid = Some(uaid);
-                            future::ok(hello_response)
-                        })
-                        .or_else(move |e| {
-                            debug!("Error registering user: {:?}", e);
-                            future::ok(err_response)
-                        }),
-                )
-            } else {
-                debug!("Deferring user registration {:?}", &uaid);
-                hello_response.uaid = Some(uaid);
-                hello_response.deferred_user_registration = Some(user);
-                future::Either::B(Box::new(future::ok(hello_response)))
-            }
-        })
-    }
+    ) -> Result<HelloResponse>;
 
-    pub fn register(
+    /// register a new channel for this UAID
+    async fn register(
```

Comment:
Can we name this:

```suggestion
    async fn register_channel(
```

and `unregister` -> `unregister_channel`